tm2/pkg/bft/consensus/state.go

```diff
@@ -1793,6 +1915,44 @@ func (cs *ConsensusState) logTelemetry(block *types.Block) {
 	metrics.BlockSizeBytes.Record(context.Background(), int64(block.Size()))
 }
 
+func (cs *ConsensusState) logErrorWithTrace(span trace.Span, msg string) {
+	if span != nil {
+		span.RecordError(errors.New(msg))
+		span.SetStatus(codes.Error, msg)
+	}
+	cs.Logger.Error(msg)
+}
+
+func (cs *ConsensusState) panicWithTrace(span trace.Span, msg string) {
+	if span != nil {
+		span.RecordError(errors.New(msg))
+		span.SetStatus(codes.Error, msg)
+	}
+	panic(msg)
+}
+
+func (cs *ConsensusState) addTrace(spanName string, opts ...trace.SpanStartOption) trace.Span {
+	var span trace.Span
+
+	if telemetry.TracesEnabled() {
+		var ctx context.Context
+		spanNameWithInfo := fmt.Sprintf("%s - CS(%v/%v/%v)", spanName, cs.Height, cs.Round, cs.Step)
+		spanNameWithInfo = strings.ReplaceAll(spanNameWithInfo, "RoundStep", "RS")
+
+		ctx, span = tracer.Start(cs.traceCtx, spanNameWithInfo, opts...)
+		if span.IsRecording() {
+			span.SetAttributes(
+				attribute.Int64("csHeight", cs.Height),
+				attribute.Int("csRound", cs.Round),
+				attribute.Int("csStep", int(cs.Step)),
+			)
+		}
+		cs.traceCtx = ctx
```

Comment:
This feels really wrong. If `addTrace` is used in the wrong place, it could create a data race. Can't we just pass the context as an argument to propagate it through the functions?

Reply:
@gfanton is right, no need to introduce `traceCtx`. The fact that context is not propagated to most of the consensus state operations is a bug, not a feature